src/Rector/Deprecation/MethodToMethodWithCheckRector.php

```diff
@@ -0,0 +1,187 @@
+     */
+    private array $configuration;
+
+    public function configure(array $configuration): void {
+        $this->configureNoticesAsComments($configuration);
+
+        foreach ($configuration as $value) {
+            if (!($value instanceof MethodToMethodWithCheckConfiguration)) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Each configuration item must be an instance of "%s"',
+                    MethodToMethodWithCheckConfiguration::class
+                ));
+            }
+        }
+
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getNodeTypes(): array {
+        return [
+            Node\Stmt\Expression::class,
+        ];
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function refactor(Node $node): ?Node {
+        assert($node instanceof Node\Stmt\Expression);
+
+        $isMethodCall = $node->expr instanceof Node\Expr\MethodCall;
+        $isAssignedMethodCall = $node->expr instanceof Node\Expr\Assign && $node->expr->expr instanceof Node\Expr\MethodCall;
+
+        if (!$isMethodCall && !$isAssignedMethodCall) {
+            return NULL;
+        }
+
+        foreach ($this->configuration as $configuration) {
+            if ($isMethodCall && $this->getName($node->expr->name) !== $configuration->getDeprecatedMethodName()) {
+                continue;
+            }
+
+            if ($isAssignedMethodCall && $this->getName($node->expr->expr->name) !== $configuration->getDeprecatedMethodName()) {
+                continue;
+            }
+
+            if ($isMethodCall) {
+                $methodNode = $this->refactorNode($node->expr, $node, $configuration);
+                if (is_null($methodNode)) {
+                    continue;
+                }
+                $node->expr = $methodNode;
+            }
+            elseif ($isAssignedMethodCall) {
+                $methodNode = $this->refactorNode($node->expr->expr, $node, $configuration);
+                if (is_null($methodNode)) {
+                    continue;
+                }
+                $node->expr->expr = $methodNode;
+            }
+
+            return $node;
+        }
+
+        return NULL;
+    }
+
+    public function refactorNode(Node\Expr\MethodCall $node, Node\Stmt\Expression $statement, MethodToMethodWithCheckConfiguration $configuration): ?Node\Expr\MethodCall {
+        $callerType = $this->nodeTypeResolver->getType($node->var);
+        $expectedType = new ObjectType($configuration->getClassName());
+
+        $isSuperOf = $expectedType->isSuperTypeOf($callerType);
+        if ($isSuperOf->yes()) {
+            $node->name = new Node\Identifier($configuration->getMethodName());
+            return $node;
+        }
+
+        if ($isSuperOf->maybe()) {
+            $node_var = $node->var->name;
+
+            if ($node->var instanceof Node\Expr\Variable) {
+                $node_var = "$$node_var";
```

Comment:
Do you know if something needs to be escaped here? Because we're doing that weird PHP thing with double `$$`.

Would this be `$node_var = "${$node_var}";`

Reply:
It is kinda weird to do something like this while we should have the power of rector and phpstan. 

Perhaps I should just try and add the failing case to a unit test, if that also fails debugging is a little easier.

Reply:
Well, it's only for th printing, and it seems to do nothing special.